pkg/clients/eks/eks.go

```diff
@@ -19,702 +19,338 @@ package eks
+// values seen in eks.Cluster.
+func LateInitialize(in *v1beta1.ClusterParameters, cluster *eks.Cluster) { // nolint:gocyclo
+	if cluster == nil {
+		return
 	}
-	versionMajor, versionMinor := versionParts[1], versionParts[2]
-
-	// reconstruct the cluster version off of major and minor
-	version := fmt.Sprintf("%v.%v", versionMajor, versionMinor)
-
-	// ami query
-	filters := map[string][]string{
-		"name":  {fmt.Sprintf("*amazon-eks-node-%v*", version)},
-		"state": {"available"},
+	if len(in.EncryptionConfig) == 0 && len(cluster.EncryptionConfig) > 0 {
+		in.EncryptionConfig = make([]v1beta1.EncryptionConfig, len(cluster.EncryptionConfig))
+		for i, e := range cluster.EncryptionConfig {
+			in.EncryptionConfig[i] = v1beta1.EncryptionConfig{
+				Resources: e.Resources,
+			}
+			if e.Provider != nil {
+				in.EncryptionConfig[i].Provider = v1beta1.Provider{
+					KeyArn: *e.Provider.KeyArn,
+				}
+			}
+		}
 	}
-
-	ec2Filters := []ec2.Filter{}
-	for name, values := range filters {
-		ec2Filters = append(ec2Filters, ec2.Filter{Name: aws.String(name), Values: values})
+	if in.Logging == nil && cluster.Logging != nil && len(cluster.Logging.ClusterLogging) > 0 {
+		setup := make([]v1beta1.LogSetup, len(cluster.Logging.ClusterLogging))
+		for i, cl := range cluster.Logging.ClusterLogging {
+			types := make([]v1beta1.LogType, len(cl.Types))
+			for j, t := range cl.Types {
+				types[j] = v1beta1.LogType(t)
+			}
+			setup[i] = v1beta1.LogSetup{
+				Enabled: cl.Enabled,
+				Types:   types,
+			}
+		}
+		in.Logging = &v1beta1.Logging{
+			ClusterLogging: setup,
+		}
 	}
+	if cluster.ResourcesVpcConfig != nil {
+		in.ResourcesVpcConfig.EndpointPrivateAccess = awsclients.LateInitializeBoolPtr(in.ResourcesVpcConfig.EndpointPrivateAccess, cluster.ResourcesVpcConfig.EndpointPrivateAccess)
+		in.ResourcesVpcConfig.EndpointPublicAccess = awsclients.LateInitializeBoolPtr(in.ResourcesVpcConfig.EndpointPublicAccess, cluster.ResourcesVpcConfig.EndpointPublicAccess)
+		if len(in.ResourcesVpcConfig.PublicAccessCidrs) == 0 && len(cluster.ResourcesVpcConfig.PublicAccessCidrs) > 0 {
+			in.ResourcesVpcConfig.PublicAccessCidrs = cluster.ResourcesVpcConfig.PublicAccessCidrs
+		}
 
-	request := e.amiClient.DescribeImagesRequest(&ec2.DescribeImagesInput{
-		Filters: ec2Filters,
-	})
-	out, err := request.Send(context.TODO())
-
-	if err != nil {
-		return nil, err
+		if len(in.ResourcesVpcConfig.SecurityGroupIDs) == 0 && len(cluster.ResourcesVpcConfig.SecurityGroupIds) > 0 {
+			in.ResourcesVpcConfig.SecurityGroupIDs = cluster.ResourcesVpcConfig.SecurityGroupIds
+		}
+		if len(in.ResourcesVpcConfig.SubnetIDs) == 0 && len(cluster.ResourcesVpcConfig.SubnetIds) > 0 {
+			in.ResourcesVpcConfig.SubnetIDs = cluster.ResourcesVpcConfig.SubnetIds
+		}
 	}
+	in.RoleArn = awsclients.LateInitializeStringPtr(in.RoleArn, cluster.RoleArn)
+	in.Version = awsclients.LateInitializeStringPtr(in.Version, cluster.Version)
+}
```

Comment:
Looks like `Tags` are not late-initialized.

Reply:
I was looking at `RDSInstance` and noticed that we don't late initialize tags there either, and I think that may be intentional. When we do the tagging in the `Initializer`, we add the tags directly to the managed resource, then they get propagated to the external infrastructure. Therefore, we will never have an empty tag map, so if we late initialize tags how are we supposed to reason about what should be there or not? Let's say a tag gets added in the AWS UI, it gets added to the managed resource by late initialization, then we want to remove it. If we explicitly remove it from the managed resource, we will hit `Observe` again, and late initialize the tag right back onto the resource, then see no diff and never make the update.

Reply:
The reason we don't late-init there was that the received object didn't have the tags, you had to make another request, as far as I remember. Late-init right now does its operation as long as the map is empty and as you pointed out this is kind of pointless since it'll never be an empty map because of the tagging initializer. But I don't really feel comfortable when one part of the codebase makes an assumption about the other on that level, i.e. they should be complete on what they do separately from each other. For example, late-init function could be used in another context or tagger could be removed etc.

Reply:
Hmm it seems like we are making the decision that Crossplane will _always_ add tags to a resource if it is an option. Are you suggesting that we check for nil tags here even though we know tags will never be nil? I get that it would be more consistent with the general purpose of late initialization, but it seems like we will always be using late initialization in this context which makes it feel extraneous.

Reply:
^ that being said, I am happy to add it if you feel strongly

Reply:
Since we can remove the tag check in `IsUpToDate` it is much cleaner if we follow your suggestion here. Thanks for catching this! :)